src/main/java/seedu/address/model/tag/UniqueTagList.java

```diff
@@ -0,0 +1,165 @@
+ * unique in terms of identity in the {@code UniqueTagList}. {@code UniqueTagList} uses {@code HashMap}
+ * to keep track of the occurrence frequency of each {@code Tag}. The {Tag} only exists in the hash map if and only
+ * if the frequency is more than 0.
+ *
+ * Supports a minimal set of list operations.
+ *
+ * @see Tag#equals(Object)
+ * @see HashMap
+ */
+public class UniqueTagList implements ReadOnlyUniqueTagList {
+
+    private final HashMap<Tag, Integer> tagFrequencyMap;
+
+    /**
+     * Constructs an {@code UniqueTagList} object.
+     */
+    public UniqueTagList() {
+        tagFrequencyMap = new HashMap<>();
+    }
+
+    /**
+     * Constructs a copy of the specified {@code UniqueTagList}.
+     *
+     * @param uniqueTagList the {@code UniqueTagList} to be copied.
+     */
+    private UniqueTagList(UniqueTagList uniqueTagList) {
+        tagFrequencyMap = new HashMap<>(uniqueTagList.tagFrequencyMap);
+    }
+
+    /**
+     * Returns a copy of this {@code UniqueTagList}.
+     *
+     * @return a copy of this {@code UniqueTagList}.
+     */
+    public UniqueTagList getCopy() {
+        return new UniqueTagList(this);
+    }
+
+    /**
+     * Adds tags in the specified tag set to this {@code UniqueTagList}.
+     * For every {@code Tag}, the frequency of the {@code Tag} will be incremented if the {@code Tag} already exists.
+     * Else, the {@code Tag} will be added to this {@code tagFrequencyMap} with a value of 1.
+     *
+     * @param tagSet the tag set containing the tags to be added.
+     */
+    public void addTags(Set<Tag> tagSet) {
+        requireNonNull(tagSet);
+
+        for (Tag tag : tagSet) {
+            requireNonNull(tag);
+            if (tagFrequencyMap.containsKey(tag)) {
+                int updatedCount = tagFrequencyMap.get(tag) + 1;
+                tagFrequencyMap.put(tag, updatedCount);
+            } else {
+                tagFrequencyMap.put(tag, 1);
+            }
+        }
+    }
+
+    /**
+     * Removes tags in the specified tag set from this {@code UniqueTagList}.
+     * For every {@code Tag}, the frequency of the {@code Tag} will be decremented if the {@code Tag} frequency is
+     * more than 1. Else, the {@code Tag} will be removed from this {@code tagFrequencyMap}.
+     *
+     * @param tagSet the tag set containing the tags to be removed.
+     */
+    public void removeTags(Set<Tag> tagSet) {
+        requireNonNull(tagSet);
+
+        for (Tag tag : tagSet) {
+            requireNonNull(tag);
+            assert tagFrequencyMap.containsKey(tag) : "Attempting to remove an un-added Tag.";
+            if (!tagFrequencyMap.containsKey(tag)) {
+                continue;
+            }
+
+            if (tagFrequencyMap.get(tag) <= 1) {
+                assert tagFrequencyMap.get(tag) == 1 : "Frequency should not be less than 1.";
+                tagFrequencyMap.remove(tag);
+            } else {
```

Comment:
Should the assertion be in the else instead? Since we want to get a warning when the code isn't going according to plan. Deleting tag when Frequency == 1 is the expected behaviour.

Reply:
In the else branch, we just need to decrement, I am just asserting if the frequency is 0 or negative. The rationale is to ensure addTags is implemented properly, e.g. if another developer add a tag into the `HashMap` with a value of less than 1. Ideally, the frequency is >= 1.